### PR TITLE
add balances and accountant for aura finance

### LINKED ISSUE #7406: Support Aura pools in gnosis chain (and more?)
## Abstract

Aura is a restaking (lke convex is to curve) for balancer pool tokens. With gnosis chain getting more traction the sDAI/EURe pool is quite heavily utilized: https://app.balancer.fi/#/gnosis-chain/pool/0xdd439304a77f54b1f7854751ac1169b279591ef7000000000000000000000064/a
https://app.aura.finance/#/100/pool/18

Would be nice to be able to decode and understand these transactions properly in gnosis chain and other chains too. Also detect balances just like we do with curve/convex.

A requirement would be to detect balancer properly in gnosis (and every chain aura is used) first: https://github.com/rotki/rotki/issues/7405

## Task

Do it

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #7406

### CODE REVIEW

File: rotkehlchen/chain/evm/decoding/aura_finance/balances.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                                method_name='balanceOf',
+                                arguments=[address],
+                            ),
+                        ) for address in addresses_with_deposits
+                    ],
+                    call_order=call_order,
+                    calls_chunk_size=chunk_size,
+                )
+                if not results:
+                    log.error(f'Failed to query aura finance balances for pool {pool_contract.address}')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/balances.py:76]: you need to add the chain name in this log

File: rotkehlchen/chain/evm/decoding/aura_finance/balances.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            counterparty=CPT_AURA_FINANCE,
+            evm_inquirer=evm_inquirer,
+            deposit_event_types={
+                (HistoryEventType.DEPOSIT, HistoryEventSubType.DEPOSIT_ASSET),
+            },
+        )
+
+    def query_balances(self) -> BalancesSheetType:
+        balances: BalancesSheetType = defaultdict(BalanceSheet)
+        addresses_with_activities = self.addresses_with_activity(event_types={
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/balances.py:44]: You should decide. Do you use `addresses_with_activity` or `addresses_with_deposits`. If you don't use the latter then you should not initialize the `deposit_event_types` in the constructor. As these type/subtype combinations are what defines what `addresses_with_deposits` returns.

Waht's worse you name the variable `addresses_with_deposits` which makes it confusing.

File: rotkehlchen/chain/evm/decoding/aura_finance/balances.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            (HistoryEventType.RECEIVE, HistoryEventSubType.RECEIVE_WRAPPED),
+        })
+        pools: set[EvmToken] = set()
+        if len(addresses_with_deposits := list(addresses_with_activities.keys())) == 0:
+            return balances
+
+        for events in addresses_with_activities.values():
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/balances.py:51]: This is the 2nd time in a row you iterate `addresses_with_activities`. Doing `list(x.keys())` also iterates it. So does in `x.values()`. Combine those iterations

File: rotkehlchen/chain/evm/decoding/aura_finance/balances.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                    if (balance := pool_contract.decode(
+                        result=result,
+                        method_name='balanceOf',
+                        arguments=[user_address],
+                    )[0]) == 0:
+                        continue
+
+                    formatted_balance = token_normalized_value_decimals(
+                        token_amount=balance,
+                        token_decimals=DEFAULT_TOKEN_DECIMALS,
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/balances.py:90]: Is it always guaranteed to be 18 decimals?

File: rotkehlchen/chain/evm/decoding/aura_finance/balances.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                        token=pool_token,
+                        inquirer=Inquirer(),
+                    )
+                    balances[user_address].assets[pool_token] += Balance(
+                        amount=formatted_balance,
+                        usd_value=formatted_balance * price,
+                    )
+            except (RemoteError, IndexError, ValueError) as e:
+                log.error(
+                    f'Failed to query {CPT_AURA_FINANCE} balances for pool {pool_token.evm_address} on '  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/balances.py:102]: no need to have a constant as part of a variable string

```suggestion
                    f'Failed to query aura balances for pool {pool_token.evm_address} on '  # noqa: E501
```

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+def query_aura_pools(evm_inquirer: 'EvmNodeInquirer') -> None:
+    """Query and store Aura Finance pools data.
+
+    May raise:
+        - RemoteError: If blockchain queries fail
+    """
+    booster_contract = EvmContract(
+        address=CHAIN_ID_TO_BOOSTER_ADDRESSES[evm_inquirer.chain_id],
+        abi=AURA_BOOSTER_ABI,
+        deployed_block=0,
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/utils.py:42]: add a comment here that this does not matter here. Same in other places you did this

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+if TYPE_CHECKING:
+    from rotkehlchen.assets.asset import EvmToken
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.inquirer import Inquirer
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+def query_aura_pools(evm_inquirer: 'EvmNodeInquirer') -> None:
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/utils.py:33]: So this is reusing a lot of code that has been abstracted by nicholas and used in places like curve lend and morpho. Check how it's done there by using `update_cached_vaults` and adjust this function to abstract it properly

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        except NotERC20Conformant as e:
+            log.error(
+                f'Failed to create/get token for address {aura_pool_token_address} '
+                f'on {evm_inquirer.chain_name}: {e!s}. Skipping...',
+            )
+            continue
+
+        if aura_pool_token.protocol != CPT_AURA_FINANCE:
+            log.debug(
+                f'Updating protocol for {evm_inquirer.chain_name} {CPT_AURA_FINANCE} '
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/utils.py:136]:
```suggestion
                f'Updating protocol for {evm_inquirer.chain_name} aura finance '
```

File: rotkehlchen/tasks/manager.py [hunk 9]
```diff
@@ -707,6 +711,29 @@ def _maybe_update_morpho_vaults(self) -> Optional[list[gevent.Greenlet]]:
...
+
+        return [self.greenlet_manager.spawn_and_track(
+                after_seconds=None,
+                task_name=f'Update Aura pools for {chain.to_name()}',
+                exception_is_error=False,
+                method=self.query_aura_pools,
+                evm_inquirer=self.chains_aggregator.get_evm_manager(chain).node_inquirer,  # type: ignore[arg-type]  # chain id is valid
+            )
+            for chain in CHAIN_ID_TO_BOOSTER_ADDRESSES
+            if should_update_protocol_cache(CacheType.AURA_POOLS, (chain.to_name(),)) is True
```
Comment (LefterisJP) [inline rotkehlchen/tasks/manager.py:734]: Question for here and in the place you save the cahce. This is not for user consumption. Can't you just use the chain id? Won't it store less bytes than the name in the DB?

File: rotkehlchen/tests/unit/test_protocol_balances.py [hunk 10]
```diff
@@ -1140,6 +1142,55 @@ def test_balancer_v2_balances(
     )
 
 
+# @pytest.mark.vcr(filter_query_parameters=['apikey'])
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_protocol_balances.py:1145]: you have commented out the VCR here.

File: rotkehlchen/tests/unit/test_protocol_balances.py [hunk 11]
```diff
@@ -1140,6 +1142,55 @@ def test_balancer_v2_balances(
...
+                    userdb=base_inquirer.database,
+                    evm_address=string_to_evm_address('0x4c42b5057a8663e2b1ac21685d1502c937a03817'),
+                    chain_id=ChainID.BASE,
+                    token_kind=EvmTokenKind.ERC20,
+                    symbol='ECLP-WETH-USDC',
+                    name='Gyroscope ECLP WETH/USDC',
+                    decimals=18,
+                ).evm_address,
+                token_kind=EvmTokenKind.ERC20,
+                weight=FVal('1'),
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_protocol_balances.py:1173]:
```suggestion
                weight=ONE,
```

File: rotkehlchen/types.py [hunk 12]
```diff
@@ -1160,6 +1160,7 @@ class CacheType(Enum):
     CURVE_LENDING_VAULT_AMM = auto()
     CURVE_LENDING_VAULT_COLLATERAL_TOKEN = auto()
     CURVE_LENDING_VAULT_BORROWED_TOKEN = auto()
+    AURA_POOLS = auto()  # stores the count of pools in db + chain
```
Comment (LefterisJP) [inline rotkehlchen/types.py:1163]: specify here what chain means (once you change it). Chain id or chain name. I would go for chain id (stringified)

File: rotkehlchen/chain/evm/decoding/morpho/utils.py [hunk 13]
```diff
@@ -40,15 +36,6 @@
...
 
-def _update_cache_timestamp(count: int | None = None) -> None:
-    with GlobalDBHandler().conn.write_ctx() as write_cursor:
-        globaldb_set_unique_cache_value(
-            write_cursor=write_cursor,
-            key_parts=[CacheType.MORPHO_VAULTS],
-            value=str(count) if count is not None else '0',
-        )
-
-
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/evm/decoding/morpho/utils.py:51]: dead code removal.

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    May raise:
+    - NotERC20Conformant
+    """
+    chain_id = vault['chain_id']
+    aura_pool_token = get_or_create_evm_token(
+        userdb=database,
+        evm_address=vault['aura_pool_token'],
+        chain_id=chain_id,
+        encounter=TokenEncounterInfo(
+            description=f'Querying {chain_id.to_name()} aura finance balances',
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/utils.py:52]: in this function you are repeating chain_id.to_name() 3 times. Put this out in a variable and reuse it

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+                    booster_contract.address,
+                    booster_contract.encode(
+                        method_name='poolInfo',
+                        arguments=[idx],
+                    ),
+                ) for idx in range(latest_pools_count)
+            ],
+        )
+    except RemoteError as e:
+        log.error(f'Failed to retrieve aura pools due to {e!s}')
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/utils.py:109]: add chain name here. It's in the inquirer

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        return None
+
+    if len(results) == 0:
+        log.debug(f'No aura pools found on {evm_inquirer.chain_name}')
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/utils.py:113]: add chain name here. It's in the inquirer

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    for idx, result in enumerate(results):
+        try:
+            pool_info = booster_contract.decode(
+                result=result,
+                method_name='poolInfo',
+                arguments=[idx],
+            )
+            pools.append({
+                'chain_id': evm_inquirer.chain_id,
+                'aura_pool_token': to_checksum_address(pool_info[3]),
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/aura_finance/utils.py:126]: don't use `to_checksum_address`. We have `deserialize_evm_address` which raises a proper `DeserializationError` which you need to catch here

File: rotkehlchen/chain/evm/decoding/aura_finance/utils.py (hunk 16 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/evm/decoding/aura_finance/utils.py:113]: it's already present in this particular diff or am i missing something?

Comment (LefterisJP) [reply rotkehlchen/chain/evm/decoding/aura_finance/utils.py:114]: you just pushed it and force pushed in the PR

**Review by LefterisJP — APPROVED** [review]
LGTM

**Pull request merged**